devito/finite_differences/coefficients.py

```diff
@@ -0,0 +1,263 @@
+        return filter_ordered((i.function for i in self.coefficients), lambda i: i.name)
+
+    @cached_property
+    def rules(self):
+
+        def generate_subs(i):
+
+            deriv_order = i.deriv_order
+            function = i.function
+            dim = i.dimension
+            weights = i.weights
+
+            fd_order = len(weights)-1
+
+            side = function.get_side(dim, deriv_order)
+            stagger = function.get_stagger(dim, deriv_order)
+
+            subs = {}
+
+            indices, x0 = generate_indices(function, dim, dim.spacing, fd_order,
+                                           side=side, stagger=stagger)
+
+            for j in range(len(weights)):
+                subs.update({function.coeff_symbol
+                             (indices[j], deriv_order, function, dim): weights[j]})
+
+            return subs
+
+        # Figure out when symbolic coefficients can be replaced
+        # with user provided coefficients and, if possible, generate
+        # replacement rules
+        rules = {}
+        for i in self.coefficients:
+            if isinstance(i.weights, np.ndarray):
+                rules.update(generate_subs(i))
+
+        return rules
+
+
+def default_rules(obj, functions):
+
+    def generate_subs(deriv_order, function, dim):
+
+        if dim.is_Time:
+            fd_order = function.time_order
+        elif dim.is_Space:
+            fd_order = function.space_order
+        else:
+            # Shouldn't arrive here
+            raise TypeError("Dimension type not recognised")
+
+        side = function.get_side(dim, deriv_order)
+        stagger = function.get_stagger(dim, deriv_order)
+
+        subs = {}
+
+        indices, x0 = generate_indices(function, dim, dim.spacing, fd_order,
+                                       side=side, stagger=stagger)
+
+        coeffs = sympy.finite_diff_weights(deriv_order, indices, x0)[-1][-1]
+
+        for j in range(len(coeffs)):
+            subs.update({function.coeff_symbol
+                         (indices[j], deriv_order, function, dim): coeffs[j]})
+
+        return subs
+
+    # Determine which 'rules' are missing
+    sym = get_sym(functions)
+    terms = obj.find(sym)
+    args_present = filter_ordered(term.args[1:] for term in terms)
+
+    subs = obj.substitutions
+    if subs:
+        args_provided = [(i.deriv_order, i.function, i.dimension)
+                         for i in subs.coefficients]
+    else:
+        args_provided = []
+
+    # NOTE: Do we want to throw a warning if the same arg has
+    # been provided twice?
+    args_provided = list(set(args_provided))
+    not_provided = list(set(args_provided).symmetric_difference(set(args_present)))
+
+    rules = {}
+    if not_provided:
+        for i in not_provided:
+            deriv_order = i[0]
+            function = i[1]
+            dim = i[2]
+            rules = {**rules, **generate_subs(deriv_order, function, dim)}
+    return rules
+
+
+def get_sym(functions):
```

Comment:
this routine makes me thing that ``coeff_symbol`` should actually be something unique and global defined somewhere here in ``coefficients.py``, rather than as a per-Function method in ``dense.py``

Reply:
I like that. This symbol is now retrieved via a function in `coefficients.py`.